### DISCUSSION TITLE
tinygrad 0.7.0

### DISCUSSION THREAD

**Discussion by wozeparrot** [tinygrad/tinygrad#1697] [Announcements]
Bigger again at 4311 lines :( But, tons of new features this time!

Just over 500 commits since `0.6.0`.

## Release Highlights

- Windows support has been dropped to focus on Linux and Mac OS.
	- Some functionality may work on Windows but no support will be provided, use WSL instead.
- [DiskTensors](https://github.com/tinygrad/tinygrad/blob/v0.7.0/tinygrad/runtime/ops_disk.py): a way to store tensors on disk has been added.
	- This is coupled with functionality in [`state.py`](https://github.com/tinygrad/tinygrad/blob/v0.7.0/tinygrad/nn/state.py) which supports saving/loading safetensors and loading torch weights.
- Tensor Cores are supported on M1/Apple Silicon and on the 7900 XTX (WMMA).
	- Support on the 7900 XTX requires weights and data to be in float16, full float16 compute support will come in a later release.
	- Tensor Core behaviour/usage is controlled by the `TC` envvar.
- Kernel optimization with nevergrad
	- This optimizes the shapes going into the kernel, gated by the `KOPT` envvar.
- P2P buffer transfers are supported on *most* AMD GPUs when using a single python process.
	- This is controlled by the `P2P` envvar.
- LLaMA 2 support.
	- A requirement of this is bfloat16 support for loading the weights, which is semi-supported by casting them to float16, proper bfloat16 support is tracked at #1290.
	- The LLaMA example now also supports 8-bit quantization using the flag `--quantize`.
- Most MLPerf models have working inference examples. Training these models is currently being worked on.
- Initial multigpu training support.
	- *slow* multigpu training by copying through host shared memory.
	- Somewhat follows torch's multiprocessing and DistributedDataParallel high-level design.
	- See the [hlb_cifar10.py](https://github.com/tinygrad/tinygrad/blob/v0.7.0/examples/hlb_cifar10.py) example.
- SymbolicShapeTracker and Symbolic JIT.
	- These two things combined allow models with changing shapes to be jitted like transformers.
	- This means that LLaMA can now be jitted for a massive increase in performance.
	- Be warned that the API for this is very WIP and may change in the future, similarly with the rest of the tinygrad API.
- [aarch64](https://github.com/tinygrad/tinygrad/blob/v0.7.0/tinygrad/renderer/assembly_arm64.py) and [ptx](https://github.com/tinygrad/tinygrad/blob/v0.7.0/tinygrad/renderer/assembly_ptx.py) assembly backend.
- WebGPU backend, see the [`compile_efficientnet.py`](https://github.com/tinygrad/tinygrad/blob/v0.7.0/examples/compile_efficientnet.py) example.
- Support for torch like tensor indexing by other tensors.
- Some more `nn` layers were promoted, namely `Embedding` and various `Conv` layers.
- [VITS](https://github.com/tinygrad/tinygrad/blob/v0.7.0/examples/vits.py) and [so-vits-svc](https://github.com/tinygrad/tinygrad/blob/v0.7.0/examples/so_vits_svc.py) examples added.
- Initial documentation work.
	- Quickstart guide: [`/docs/quickstart.md`](https://github.com/tinygrad/tinygrad/blob/v0.7.0/docs/quickstart.md)
	- Environment variable reference: [`/docs/env_vars.md`](https://github.com/tinygrad/tinygrad/blob/v0.7.0/docs/env_vars.md)

And lots of small optimizations all over the codebase.

### Join the [Discord](https://discord.gg/beYbxwxVdx)!

<hr /><em>This discussion was created from the release <a href='https://github.com/tinygrad/tinygrad/releases/tag/v0.7.0'>tinygrad 0.7.0</a>.</em>

**Comment by Demonese** [comment]
Windows support: R.I.P

**Reply by geohot** [reply]
If someone wants to sponsor Windows at $1,000/mo, we'll put Windows back.

**Reply by Demonese** [reply]
try WSL2

**Comment by rsokl** [comment]
Your VITS and docs links are broken.

**Comment by geohot** [comment]
How'd you get 4942 lines?

git checkout v0.7.0 && ./sz.py

I see 4311

**Reply by wozeparrot** [reply]
~~Oh, I ran sz.sh, wasn't sure which one should have been used, maybe we should drop the other one if sz.py is the correct one?~~

**Reply by geohot** [reply]
Where is sz.sh, I don't even have that one.

**Reply by wozeparrot** [reply]
Ohhh that's my bad, the way i sync my repo doesn't delete files, so the sz.sh to sz.py changr didnt delete sz.sh.

**Comment by sabintamang1** [comment]
obs appears to be running a ChromeOS container. unsupported
